File: ios/MeridianApp/Core/ScheduleTimeParser.swift
// ScheduleTimeParser.swift
// Meridian — iOS 27 / Swift 6
//
// Deterministically resolves a natural-language date/time phrase to a `Date`
// in a given zone, anchored at "now". Ported 1:1 from Android's
// `core/time/ScheduleTimeParser.kt`.
//
// This is the app's own time math — the LLM only *extracts* the phrase
// ("next Monday 2pm", "tomorrow", "2026-06-22 14:00") and this parser turns it
// into an exact instant, so a weak on-device model can never pick the wrong time.
//
// Supported, case-insensitive:
//  - absolute ISO dates `yyyy-MM-dd` (optionally with a time);
//  - relative days: today/tonight, tomorrow, day after tomorrow, yesterday;
//  - weekdays with optional `this`/`next` (e.g. "next friday"), and "next week";
//  - month-day: "June 22", "22 Jun" (year optional; rolls to next year if past);
//  - durations from now: "in 30 minutes", "in 2 hours", "in a week";
//  - times: "2pm", "2:30 pm", 24-hour "14:00", noon/midnight/morning/etc.
//
// Conventions: a bare or "this" weekday is the soonest occurrence *including
// today*; "next" weekday skips today. A date with no time defaults to 09:00; a
// time with no date is today if still ahead, otherwise tomorrow. Returns nil if
// nothing parses.

import Foundation

enum ScheduleTimeParser {

    /// Default time-of-day applied to a date that arrives without one (09:00).
    static let defaultHour = 9
    static let defaultMinute = 0

    /// Relative durations beyond this many years out are treated as unparseable.
    private static let maxFutureYears = 200

    // MARK: - Public entry point

    /// Resolves `phrase` to an absolute `Date` in `zone`, anchored at `now`.
    static func parse(_ phrase: String, zone: TimeZone, now: Date) -> Date? {
        let text = phrase.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty { return nil }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone

        let time = parseTimeOfDay(text)

        // "in N <unit>" sets a full instant; for day/week units an explicit
        // time-of-day still applies.
        if let relative = parseRelativeDuration(text, calendar: calendar, now: now, time: time) {
            return relative
        }

        let today = calendar.dateComponents([.year, .month, .day], from: now)
        guard let todayDate = calendar.date(from: today) else { return nil }
        let date = parseDate(text, today: todayDate, calendar: calendar)

        // Build the local wall-clock date/time, then convert to an absolute Date.
        if let date {
            let t = time ?? (hour: defaultHour, minute: defaultMinute)
            return makeDate(date: date, hour: t.hour, minute: t.minute, calendar: calendar)
        }
        if let time {
            // Today at `time` if still ahead of `now`, otherwise tomorrow.
            guard let todayAt = makeDate(date: todayDate, hour: time.hour, minute: time.minute, calendar: calendar)
            else { return nil }
            if todayAt > now { return todayAt }
            guard let tomorrow = calendar.date(byAdding: .day, value: 1, to: todayDate) else { return nil }
            return makeDate(date: tomorrow, hour: time.hour, minute: time.minute, calendar: calendar)
        }
        return nil
    }

    // MARK: - Relative durations

    private static func parseRelativeDuration(
        _ text: String,
        calendar: Calendar,
        now: Date,
        time: (hour: Int, minute: Int)?
    ) -> Date? {
        guard let match = firstMatch(relativeDurationPattern, in: text) else { return nil }
        let countRaw = match[1]
        let unit = match[2]

        let n: Int
        if countRaw == "a" || countRaw == "an" {
            n = 1
        } else if let parsed = Int(countRaw) {
            n = parsed
        } else {
            return nil
        }

        let result: Date?
        if unit.hasPrefix("min") {
            // Minutes/hours are absolute offsets from now; a stray time-of-day doesn't apply.
            result = calendar.date(byAdding: .minute, value: n, to: now)
        } else if unit.hasPrefix("h") {
            result = calendar.date(byAdding: .hour, value: n, to: now)
        } else if unit.hasPrefix("day") {
            result = applyTime(calendar.date(byAdding: .day, value: n, to: now), time, calendar: calendar)
        } else if unit.hasPrefix("w") {
            result = applyTime(calendar.date(byAdding: .day, value: n * 7, to: now), time, calendar: calendar)
        } else {
            return nil
        }

        guard let result else { return nil }
        // Treat anything past a sane horizon as unparseable rather than scheduling it.
        let resultYear = calendar.component(.year, from: result)
        let nowYear = calendar.component(.year, from: now)
        if resultYear > nowYear + maxFutureYears { return nil }
        return result
    }

    /// Replaces the time-of-day of `date` with `time` (keeping its calendar day), or returns `date`.
    private static func applyTime(
        _ date: Date?,
        _ time: (hour: Int, minute: Int)?,
        calendar: Calendar
    ) -> Date? {
        guard let date else { return nil }
        guard let time else { return date }
        var comps = calendar.dateComponents([.year, .month, .day], from: date)
        comps.hour = time.hour
        comps.minute = time.minute
        comps.second = 0
        return calendar.date(from: comps)
    }

    // MARK: - Time of day

    private static func parseTimeOfDay(_ text: String) -> (hour: Int, minute: Int)? {
        if let m = firstMatch(amPmTimePattern, in: text) {
            var h = Int(m[1]) ?? 0
            let min = Int(m[2]) ?? 0
            let pm = m[3].replacingOccurrences(of: ".", with: "") == "pm"
            if (1...12).contains(h) && (0...59).contains(min) {
                if h == 12 { h = 0 }
                if pm { h += 12 }
                return (h, min)
            }
        }
        if let m = firstMatch(twentyFourHourPattern, in: text) {
            let h = Int(m[1]) ?? -1
            let min = Int(m[2]) ?? -1
            if (0...23).contains(h) && (0...59).contains(min) { return (h, min) }
        }
        if text.contains("midnight") { return (0, 0) }
        // Word-boundary match: "noon" must not fire inside "afternoon".
        if containsMatch(noonPattern, in: text) { return (12, 0) }
        if text.contains("afternoon") { return (14, 0) }
        if text.contains("morning") { return (9, 0) }
        if text.contains("evening") { return (18, 0) }
        if text.contains("tonight") || text.contains("night") { return (20, 0) }
        return nil
    }

    // MARK: - Date

    private static func parseDate(_ text: String, today: Date, calendar: Calendar) -> Date? {
        if let m = firstMatch(isoDatePattern, in: text),
           let year = Int(m[1]), let month = Int(m[2]), let day = Int(m[3]) {
            var comps = DateComponents()
            comps.year = year; comps.month = month; comps.day = day
            comps.hour = 0; comps.minute = 0; comps.second = 0
            if let d = calendar.date(from: comps),
               calendar.component(.month, from: d) == month,
               calendar.component(.day, from: d) == day {
                return d
            }
        }

        if text.contains("day after tomorrow") { return calendar.date(byAdding: .day, value: 2, to: today) }
        if text.contains("tomorrow") { return calendar.date(byAdding: .day, value: 1, to: today) }
        if text.contains("today") || text.contains("tonight") { return today }
        if text.contains("yesterday") { return calendar.date(byAdding: .day, value: -1, to: today) }

        // "next week [monday]" anchors the weekday search a week ahead.
        let nextWeek = containsMatch(nextWeekPattern, in: text)
        let anchor = nextWeek ? (calendar.date(byAdding: .day, value: 7, to: today) ?? today) : today
        if let weekday = parseWeekday(text, anchor: anchor, calendar: calendar) { return weekday }
        if nextWeek { return anchor }
        if let monthDay = parseMonthDay(text, today: today, calendar: calendar) { return monthDay }
        return nil
    }

    private static func parseWeekday(_ text: String, anchor: Date, calendar: Calendar) -> Date? {
        for (name, dow) in weekdays {
            guard containsMatch("\\b\(name)\\b", in: text) else { continue }
            let isNext = containsMatch("\\bnext\\s+\(name)\\b", in: text)
            // Gregorian: Calendar weekday 1 = Sunday ... 7 = Saturday.
            // `dow` here is 1 = Monday ... 7 = Sunday (java.time semantics); convert.
            let anchorWeekday = isoWeekday(of: anchor, calendar: calendar)
            var days = ((dow - anchorWeekday) % 7 + 7) % 7
            if isNext && days == 0 { days = 7 }
            return calendar.date(byAdding: .day, value: days, to: anchor)
        }
        return nil
    }

    private static func parseMonthDay(_ text: String, today: Date, calendar: Calendar) -> Date? {
        for (name, month) in months {
            let after = firstMatch("\\b\(name)\\b\\.?\\s+(\\d{1,2})(?:st|nd|rd|th)?(?:,?\\s+(\\d{4}))?\\b", in: text)
            let before = firstMatch("\\b(\\d{1,2})(?:st|nd|rd|th)?\\s+\(name)\\b(?:,?\\s+(\\d{4}))?", in: text)
            guard let match = after ?? before else { continue }
            guard let day = Int(match[1]) else { continue }
            let year = match.count > 2 ? Int(match[2]) : nil

            let todayYear = calendar.component(.year, from: today)
            var comps = DateComponents()
            comps.year = year ?? todayYear
            comps.month = month
            comps.day = day
            comps.hour = 0; comps.minute = 0; comps.second = 0
            guard var date = calendar.date(from: comps),
                  calendar.component(.month, from: date) == month,
                  calendar.component(.day, from: date) == day
            else { continue }
            // No explicit year and the date already passed this year → next year.
            if year == nil && date < today {
                date = calendar.date(byAdding: .year, value: 1, to: date) ?? date
            }
            return date
        }
        return nil
    }

    // MARK: - Helpers

    /// Builds an absolute `Date` from a calendar-day `date` plus an explicit hour/minute.
    private static func makeDate(date: Date, hour: Int, minute: Int, calendar: Calendar) -> Date? {
        var comps = calendar.dateComponents([.year, .month, .day], from: date)
        comps.hour = hour
        comps.minute = minute
        comps.second = 0
        return calendar.date(from: comps)
    }

    /// ISO weekday number for `date`: 1 = Monday ... 7 = Sunday (matches `weekdays` table).
    private static func isoWeekday(of date: Date, calendar: Calendar) -> Int {
        // Calendar.weekday: 1 = Sunday ... 7 = Saturday. Convert to ISO Monday=1.
        let w = calendar.component(.weekday, from: date)
        return ((w + 5) % 7) + 1
    }

    // MARK: - Lookup tables (ISO weekday: Monday = 1 ... Sunday = 7)

    private static let weekdays: [(String, Int)] = [
        ("monday", 1), ("tuesday", 2), ("wednesday", 3),
        ("thursday", 4), ("friday", 5), ("saturday", 6), ("sunday", 7),
        ("mon", 1), ("tue", 2), ("wed", 3),
        ("thu", 4), ("fri", 5), ("sat", 6), ("sun", 7),
    ]

    private static let months: [(String, Int)] = [
        ("january", 1), ("february", 2), ("march", 3), ("april", 4),
        ("may", 5), ("june", 6), ("july", 7), ("august", 8),
        ("september", 9), ("october", 10), ("november", 11), ("december", 12),
        ("jan", 1), ("feb", 2), ("mar", 3), ("apr", 4),
        ("jun", 6), ("jul", 7), ("aug", 8), ("sep", 9),
        ("sept", 9), ("oct", 10), ("nov", 11), ("dec", 12),
    ]

    // MARK: - Regex patterns

    private static let relativeDurationPattern =
        "\\bin\\s+(\\d+|an?)\\s+(minutes?|mins?|hours?|hrs?|days?|weeks?)\\b"
    private static let amPmTimePattern =
        "(?<!\\d)(\\d{1,2})(?::(\\d{2}))?\\s*(a\\.?m\\.?|p\\.?m\\.?)(?![a-z])"
    // Digit lookarounds (not \b) so an ISO "…-22T14:00" still yields the 14:00 time.
    private static let twentyFourHourPattern = "(?<![\\d:])(\\d{1,2}):(\\d{2})(?!\\d)"
    private static let isoDatePattern = "(?<!\\d)(\\d{4})-(\\d{1,2})-(\\d{1,2})(?!\\d)"
    private static let nextWeekPattern = "\\bnext\\s+week\\b"
    private static let noonPattern = "\\b(noon|midday)\\b"

    // MARK: - Regex utilities

    /// Returns the capture groups of the first match (index 0 = whole match, 1.. = groups).
    /// Missing optional groups are returned as empty strings.
    private static func firstMatch(_ pattern: String, in text: String) -> [String]? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return nil
        }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        guard let match = regex.firstMatch(in: text, options: [], range: range) else { return nil }
        var groups: [String] = []
        for i in 0..<match.numberOfRanges {
            if let r = Range(match.range(at: i), in: text) {
                groups.append(String(text[r]))
            } else {
                groups.append("")
            }
        }
        return groups
    }

    private static func containsMatch(_ pattern: String, in text: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return false
        }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return regex.firstMatch(in: text, options: [], range: range) != nil
    }
}

File: app/src/main/java/com/example/core/time/ScheduleTimeParser.kt
package com.example.core.time

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.Month
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.Locale

/**
 * Deterministically resolves a natural-language date/time phrase to an [Instant] in a given zone,
 * anchored at "now". This is the app's own time math (§12.7 — the LLM never computes time): the
 * assistant only *extracts* the phrase ("next Monday 2pm", "tomorrow", "2026-06-22 14:00") and this
 * parser turns it into an exact instant, so a weak on-device model can no longer pick the wrong time.
 *
 * Supported, case-insensitive:
 *  - absolute ISO dates `yyyy-MM-dd` (optionally with a time);
 *  - relative days: today/tonight, tomorrow, day after tomorrow, yesterday;
 *  - weekdays with optional `this`/`next` (e.g. "next friday"), and "next week";
 *  - month-day: "June 22", "22 Jun" (year optional; rolls to next year if already past);
 *  - durations from now: "in 30 minutes", "in 2 hours", "in a week";
 *  - times: "2pm", "2:30 pm", 24-hour "14:00", and noon/midnight/morning/afternoon/evening/tonight.
 *
 * Conventions: a bare or "this" weekday is the soonest occurrence *including today*; "next" weekday
 * skips today (so on a Monday, "next Monday" is +7). A date with no time defaults to [DEFAULT_TIME];
 * a time with no date is today if still ahead, otherwise tomorrow. Returns null if nothing parses.
 */
object ScheduleTimeParser {

    val DEFAULT_TIME: LocalTime = LocalTime.of(9, 0)

    /** Relative durations beyond this many years out are treated as unparseable, not scheduled. */
    private const val MAX_FUTURE_YEARS = 200

    fun parse(phrase: String, zone: ZoneId, now: Instant): Instant? {
        val text = phrase.lowercase(Locale.ROOT).trim()
        if (text.isEmpty()) return null
        val nowZdt = ZonedDateTime.ofInstant(now, zone)
        val time = parseTimeOfDay(text)

        // "in N <unit>" sets a full instant; for day/week units an explicit time-of-day still applies
        // ("in 2 days at 3pm" → that date at 15:00, not at the current wall-clock time).
        parseRelativeDuration(text, nowZdt, time)?.let { return it.toInstant() }

        val date = parseDate(text, nowZdt.toLocalDate())

        // A date plus an *explicit but impossible* clock time ("tomorrow 25:00", "…T24:00",
        // "13pm") is a user mistake, not an absent time: fail closed instead of silently booking
        // the 09:00 default nine hours away from what was asked.
        if (date != null && time == null && hasMalformedExplicitTime(text)) return null

        val local = when {
            date != null && time != null -> date.atTime(time)
            date != null -> date.atTime(DEFAULT_TIME)
            time != null -> {
                val todayAt = nowZdt.toLocalDate().atTime(time)
                if (todayAt.atZone(zone).toInstant().isAfter(now)) todayAt else todayAt.plusDays(1)
            }
            else -> return null
        }
        return local.atZone(zone).toInstant()
    }

    private fun parseRelativeDuration(text: String, nowZdt: ZonedDateTime, time: LocalTime?): ZonedDateTime? {
        val m = RELATIVE_DURATION.find(text) ?: return null
        val n = m.groupValues[1].let { if (it == "a" || it == "an") 1L else it.toLongOrNull() ?: return null }
        val unit = m.groupValues[2]
        // Absurd counts ("in 9999999999 weeks") overflow or land hundreds of millions of years out —
        // treat anything past a sane horizon (or that throws) as unparseable rather than scheduling it.
        return runCatching {
            val result: ZonedDateTime = when {
                // Minutes/hours are absolute offsets from now; a stray time-of-day doesn't apply.
                unit.startsWith("min") -> nowZdt.plusMinutes(n)
                unit.startsWith("h") -> nowZdt.plusHours(n)
                // Day/week offsets land on a date — honor an explicit time if the phrase had one.
                unit.startsWith("day") -> nowZdt.plusDays(n).applyTime(time)
                unit.startsWith("w") -> nowZdt.plusWeeks(n).applyTime(time)
                else -> return@runCatching null
            }
            if (result.year > nowZdt.year + MAX_FUTURE_YEARS) null else result
        }.getOrNull()
    }

    private fun ZonedDateTime.applyTime(time: LocalTime?): ZonedDateTime =
        if (time == null) this else toLocalDate().atTime(time).atZone(zone)

    /**
     * True when the phrase contains an explicit numeric clock candidate that failed validation —
     * hours/minutes out of range in either 24-hour or am/pm form. Distinguishes "no time given"
     * from "given but unparseable" so callers can reject instead of defaulting.
     */
    private fun hasMalformedExplicitTime(text: String): Boolean {
        ATTACHED_CLOCK_TIME.containsMatchIn(text) && return true
        ATTACHED_AM_PM_TIME.containsMatchIn(text) && return true
        AM_PM_TIME.find(text)?.let { m ->
            val h = m.groupValues[1].toInt()
            val min = m.groupValues[2].toIntOrNull() ?: 0
            if (h !in 1..12 || min !in 0..59) return true
        }
        TWENTY_FOUR_HOUR.find(text)?.let { m ->
            val h = m.groupValues[1].toInt()
            val min = m.groupValues[2].toInt()
            if (h !in 0..23 || min !in 0..59) return true
        }
        return false
    }

    private fun parseTimeOfDay(text: String): LocalTime? {        AM_PM_TIME.find(text)?.let { m ->
            var h = m.groupValues[1].toInt()
            val min = m.groupValues[2].toIntOrNull() ?: 0
            val pm = m.groupValues[3].replace(".", "") == "pm"
            if (h in 1..12 && min in 0..59) {
                if (h == 12) h = 0
                if (pm) h += 12
                return LocalTime.of(h, min)
            }
        }
        TWENTY_FOUR_HOUR.find(text)?.let { m ->
            val h = m.groupValues[1].toInt()
            val min = m.groupValues[2].toInt()
            if (h in 0..23 && min in 0..59) return LocalTime.of(h, min)
        }
        return when {
            "midnight" in text -> LocalTime.MIDNIGHT
            // Word-boundary match: "noon" must not fire on the "noon" inside "afternoon".
            NOON.containsMatchIn(text) -> LocalTime.NOON
            "afternoon" in text -> LocalTime.of(14, 0)
            "morning" in text -> LocalTime.of(9, 0)
            "evening" in text -> LocalTime.of(18, 0)
            "tonight" in text || "night" in text -> LocalTime.of(20, 0)
            else -> null
        }
    }

    private fun parseDate(text: String, today: LocalDate): LocalDate? {
        ISO_DATE.find(text)?.let { m ->
            runCatching {
                LocalDate.of(m.groupValues[1].toInt(), m.groupValues[2].toInt(), m.groupValues[3].toInt())
            }.getOrNull()?.let { return it }
        }
        when {
            "day after tomorrow" in text -> return today.plusDays(2)
            "tomorrow" in text -> return today.plusDays(1)
            "today" in text || "tonight" in text -> return today
            "yesterday" in text -> return today.minusDays(1)
        }
        // "next week [monday]" anchors the weekday search a week ahead (so it isn't this week's).
        val nextWeek = NEXT_WEEK.containsMatchIn(text)
        val anchor = if (nextWeek) today.plusWeeks(1) else today
        parseWeekday(text, anchor)?.let { return it }
        if (nextWeek) return anchor
        parseMonthDay(text, today)?.let { return it }
        return null
    }

    private fun parseWeekday(text: String, anchor: LocalDate): LocalDate? {
        for ((name, dow) in WEEKDAYS) {
            if (!Regex("\\b$name\\b").containsMatchIn(text)) continue
            val isNext = Regex("\\bnext\\s+$name\\b").containsMatchIn(text)
            var days = ((dow.value - anchor.dayOfWeek.value) % 7 + 7) % 7
            if (isNext && days == 0) days = 7
            return anchor.plusDays(days.toLong())
        }
        return null
    }

    private fun parseMonthDay(text: String, today: LocalDate): LocalDate? {
        for ((name, month) in MONTHS) {
            val after = Regex("\\b$name\\b\\.?\\s+(\\d{1,2})(?:st|nd|rd|th)?(?:,?\\s+(\\d{4}))?\\b").find(text)
            val before = Regex("\\b(\\d{1,2})(?:st|nd|rd|th)?\\s+$name\\b(?:,?\\s+(\\d{4}))?").find(text)
            val match = after ?: before ?: continue
            val day = match.groupValues[1].toIntOrNull() ?: continue
            val year = match.groupValues[2].toIntOrNull()
            var date = runCatching { LocalDate.of(year ?: today.year, month.value, day) }.getOrNull() ?: continue
            // No explicit year and the date already passed this year → the user means next year.
            if (year == null && date.isBefore(today)) date = date.plusYears(1)
            return date
        }
        return null
    }

    private val WEEKDAYS: List<Pair<String, DayOfWeek>> = listOf(
        "monday" to DayOfWeek.MONDAY, "tuesday" to DayOfWeek.TUESDAY, "wednesday" to DayOfWeek.WEDNESDAY,
        "thursday" to DayOfWeek.THURSDAY, "friday" to DayOfWeek.FRIDAY, "saturday" to DayOfWeek.SATURDAY,
        "sunday" to DayOfWeek.SUNDAY,
        "mon" to DayOfWeek.MONDAY, "tue" to DayOfWeek.TUESDAY, "wed" to DayOfWeek.WEDNESDAY,
        "thu" to DayOfWeek.THURSDAY, "fri" to DayOfWeek.FRIDAY, "sat" to DayOfWeek.SATURDAY,
        "sun" to DayOfWeek.SUNDAY,
    )

    private val MONTHS: List<Pair<String, Month>> = listOf(
        "january" to Month.JANUARY, "february" to Month.FEBRUARY, "march" to Month.MARCH,
        "april" to Month.APRIL, "may" to Month.MAY, "june" to Month.JUNE, "july" to Month.JULY,
        "august" to Month.AUGUST, "september" to Month.SEPTEMBER, "october" to Month.OCTOBER,
        "november" to Month.NOVEMBER, "december" to Month.DECEMBER,
        "jan" to Month.JANUARY, "feb" to Month.FEBRUARY, "mar" to Month.MARCH, "apr" to Month.APRIL,
        "jun" to Month.JUNE, "jul" to Month.JULY, "aug" to Month.AUGUST, "sep" to Month.SEPTEMBER,
        "sept" to Month.SEPTEMBER, "oct" to Month.OCTOBER, "nov" to Month.NOVEMBER, "dec" to Month.DECEMBER,
    )

    private val RELATIVE_DURATION =
        Regex("\\bin\\s+(\\d+|an?)\\s+(minutes?|mins?|hours?|hrs?|days?|weeks?)\\b")
    // Sign characters are excluded from both lookbehinds so "-1:30"/"-3pm" are never read as
    // plain "1:30"/"3pm" — a negative clock time is a user mistake, not a valid time of day.
    private val AM_PM_TIME = Regex("(?<![\\d+\\-])(\\d{1,2})(?::(\\d{2}))?\\s*(a\\.?m\\.?|p\\.?m\\.?)(?![a-z])")
    // Digit lookarounds (not \b) so an ISO "…-22T14:00" still yields the date and the 14:00 time —
    // a letter like 'T' next to a digit is not a word boundary.
    private val TWENTY_FOUR_HOUR = Regex("(?<![\\d:+\\-])(\\d{1,2}):(\\d{2})(?!\\d)")
    private val ATTACHED_CLOCK_TIME = Regex("(?<=[+\\-/])\\s*(\\d{1,2}):(\\d{2})(?!\\d)")
    private val ATTACHED_AM_PM_TIME = Regex("(?<=[+\\-/])\\s*(\\d{1,2})\\s*(?:a\\.?m\\.?|p\\.?m\\.?)")
    private val ISO_DATE = Regex("(?<!\\d)(\\d{4})-(\\d{1,2})-(\\d{1,2})(?!\\d)")
    private val NEXT_WEEK = Regex("\\bnext\\s+week\\b")
    private val NOON = Regex("\\b(noon|midday)\\b")
}
